Mathlib/Analysis/SpecialFunctions/Log/NegMulLog.lean

```diff
@@ -52,18 +52,128 @@ lemma hasDerivAt_mul_log {x : ℝ} (hx : x ≠ 0) : HasDerivAt (fun x ↦ x * lo
   refine DifferentiableOn.differentiableAt differentiableOn_mul_log ?_
   simp [hx]
 
-lemma deriv2_mul_log {x : ℝ} (hx : x ≠ 0) : deriv^[2] (fun x ↦ x * log x) x = x⁻¹ := by
+open Filter in
+lemma tendsto_deriv_mul_log_nhdsWithin_zero :
```

Comment:
@dupuisf said (https://github.com/leanprover-community/mathlib4/pull/9734#discussion_r1669489433)

    This can be moved up to just below hasDerivAt_mul_log without changing anything. And from this, you can probably prove lemma isLittleO_nhds_zero_id_deriv_mul_log : id =o[𝓝[>] 0] (deriv (fun x ↦ x * log x)) fairly directly, and then use that to prove non-differentiability faster.

I currently know neither how to prove it (directly) nor how to use it to prove non-differentiability faster. Will have to look at what lemmas can be used to connect those things...

Reply:
On second thought, going through `IsLittleO` is probably not the best way. I think we need lemmas that say that if `deriv f` tendsto `atTop` (or `atBot`) in the neighborhood of some point, then `f` is not differentiable at that point. This could probably go in `MeanValue.lean`. With this it should follow directly from that lemma. Unfortunately we have very few lemmas saying that a function is *not* differentiable...

I could have a crack at it in the next few days if you've had enough!

Reply:
> I could have a crack at it in the next few days if you've had enough!

Sure, go ahead. Even if I could probably prove those things given some time, I certainly wouldn't get the API right on the first few tries.

Reply:
Alright, these two lemmas should do it (with `atBot` and `𝓝[<] a)` versions):
```lean
theorem ContinuousWithinAt.mono_nhdsWithin  {α : Type u_1} {β : Type u_2} [TopologicalSpace α]
    [TopologicalSpace β] {f : α → β} {s₁ s₂ : Set α} {x : α}
    (h : ContinuousWithinAt f s₁ x) (h_nhds : 𝓝[s₂] x ≤ 𝓝[s₁] x) : ContinuousWithinAt f s₂ x :=
  calc (𝓝[s₂] x).map f ≤ (𝓝[s₁] x).map f := by gcongr
    _ ≤ 𝓝 (f x) := h

theorem not_differentiableAt_of_deriv_tendsto_atTop_Ioi (f : ℝ → ℝ) {a : ℝ}
    (hf : Tendsto (deriv f) (𝓝[>] a) atTop) : ¬ DifferentiableAt ℝ f a := by
  by_cases hcont_at_a : ¬ ContinuousWithinAt f (Ici a) a
  · exact fun hcontra => hcont_at_a <| hcontra.continuousAt.continuousWithinAt
  · push_neg at hcont_at_a
    intro hdiff
    have hdiff₂ := hdiff.hasDerivAt
    rw [hasDerivAt_iff_tendsto_slope] at hdiff₂
    replace hdiff₂ : Tendsto (slope f a) (𝓝[>] a) (𝓝 (deriv f a)) :=
      hdiff₂.mono_left <| nhds_right'_le_nhds_ne a
    have h₁ : ∀ᶠ b in (𝓝[>] a), ∀ x ∈ Ioc a b, deriv f a + 1 < deriv f x := by
      rw [(nhdsWithin_Ioi_basis a).eventually_iff]
      rw [(nhdsWithin_Ioi_basis a).tendsto_left_iff] at hf
      obtain ⟨b, hab, hb⟩ := hf (Ioi (deriv f a + 1)) (Ioi_mem_atTop _)
      refine ⟨b, hab, fun x hx z hz => ?_⟩
      simp only [MapsTo, mem_Ioo, mem_Ioi, and_imp] at hb
      exact hb hz.1 <| hz.2.trans_lt hx.2
    have h₂ : ∀ᶠ b in (𝓝[>] a), slope f a b < (deriv f a) + (1 : ℝ) := by
      rw [(nhds_basis_Ioo _).tendsto_right_iff] at hdiff₂
      specialize hdiff₂ ⟨deriv f a - 1, deriv f a + 1⟩ <| by simp
      filter_upwards [hdiff₂] with z hz using hz.2
    have h₃ : ∀ᶠ b in (𝓝[>] a), ∀ x ∈ Ioc a b, 0 < deriv f x := by
      rw [(nhdsWithin_Ioi_basis a).eventually_iff]
      rw [(nhdsWithin_Ioi_basis a).tendsto_left_iff] at hf
      obtain ⟨b, hab, hb⟩ := hf (Ioi 0) (Ioi_mem_atTop 0)
      refine ⟨b, hab, fun x hx z hz => ?_⟩
      simp only [MapsTo, mem_Ioo, mem_Ioi, and_imp] at hb
      exact hb hz.1 <| hz.2.trans_lt hx.2
    have hcontra : ∀ᶠ _ in (𝓝[>] a), False := by
      filter_upwards [h₁, h₂, h₃, eventually_mem_nhdsWithin] with
        b hb hslope hderiv_pos (hab : a < b)
      have hdiff' : DifferentiableOn ℝ f (Ioc a b) := fun z hz => by
        have : deriv f z ≠ 0 := Ne.symm <| ne_of_lt <| hderiv_pos z hz
        exact DifferentiableAt.differentiableWithinAt <| differentiableAt_of_deriv_ne_zero this
      have hcont : ContinuousOn f (Icc a b) := by
        intro z hz
        by_cases hz' : z = a
        · rw [hz']
          exact hcont_at_a.mono Icc_subset_Ici_self
        · have hz'' : z ∈ Ioc a b := ⟨lt_of_le_of_ne hz.1 (Ne.symm hz'), hz.2⟩
          refine (hdiff'.continuousOn z hz'').mono_nhdsWithin ?_
          by_cases hz''' : z = b
          · rw [hz''', nhdsWithin_Ioc_eq_nhdsWithin_Iic hab, nhdsWithin_Icc_eq_nhdsWithin_Iic hab]
          · have hfinal₁ : 𝓝[Ioc a b] z = 𝓝 z := by
              rw [nhdsWithin_eq_nhds]
              refine Ioc_mem_nhds ?_ (lt_of_le_of_ne hz.2 hz''')
              simp_all only [mem_Ioc, and_imp, mem_Icc, and_true]
            have hfinal₂ : 𝓝[Icc a b] z = 𝓝 z := by
              rw [nhdsWithin_eq_nhds]
              exact Icc_mem_nhds (lt_of_le_of_ne hz.1 (ne_of_lt hz''.1)) (lt_of_le_of_ne hz.2 hz''')
            rw [hfinal₁, hfinal₂]
      obtain ⟨x, hx₁, hx₂⟩ := exists_deriv_eq_slope' f hab hcont (hdiff'.mono Ioo_subset_Ioc_self)
      specialize hb x (mem_Ioc_of_Ioo hx₁)
      rw [hx₂] at hb
      linarith
    simp [Filter.eventually_false_iff_eq_bot, ← not_mem_closure_iff_nhdsWithin_eq_bot] at hcontra
```
Do you want to add these to this PR, or would you prefer it if I PRed them separately?

Reply:
Wow, that's a much bigger proof than I expected. Maybe I wouldn't have found it after all... Thanks! 

Is `exists_deriv_eq_slope'` missing? Maybe you could either push to this branch or make a new PR?

Reply:
Indeed `exists_deriv_eq_slope'` is missing from that code snippet (it's the same as `exists_deriv_eq_slope` but it actually uses `slope`). I'll just open a separate PR, I think it's a bit easier (also that means I'm not too involved in this one to finish reviewing it). Once that is merged, you can just merge master into this one.

Reply:
Here's the PR: #15200

Reply:
Used this lemma. Removed auxiliary stuff.